Added support for UIColor

sgl0v opened the pull request:
Implemented feature to edit color value with RGB/HSB color pickers.

lightice11 commented:
Can we get an update on this @grp ? It sucks to neglect pull requests that fix issues that you have created (#5).

andyscott commented:
I'd also love to see this merged.

todaylover commented:
anyone cares this issue?

facebook-github-bot-2 commented:
Thank you for reporting this issue and appreciate your patience. We've  notified the core team for an update on this issue. We're looking for a response within the next 30 days or the issue may be closed.

sgl0v commented:
Hello, guys. Finally I've updated the pull request(merged master branch & fixed all issues I had) and made some changes in the UI (now it is UITableView-based and looks consistent with all other components.). Enjoy!

grp commented:
I took a quick look at it looks great! A few overall points before I add some notes inline:

 - Interesting for the choice of representing a color as a hex-string — any reason not to use UIColor directly? Since the 2.0.0 update (adding array support), you should be able to put any NSObject into the tweak value field since they're created lazily through a block at runtime. This would also help avoiding type issues if you legitimately do want a Tweak with a string starting with "#".
 - A few of the views use auto-layout, but we generally don't use that at Facebook and it occasionally causes compatibility issues. Could you make those use `-layoutSubviews` instead? (If that would be complicated, it's probably fine as-is, but a little harder to understand.)
 - Mostly a nit: the rest of the code has the `*` in pointers next to the variable name, rather than next to the type as you have it for your changes.

grp commented:
Finally, the HSV support is pretty cool, but it looks like it adds a few more classes just for that. Maybe for simplicity it could just be RGB to start?

grp commented on FBTweak/FBColorUtils.h:
```diff
@@ -0,0 +1,80 @@
+ *  @param rgb   The rgb color values
+ *  @param outHSB The hsb color values
+ */
+extern void FBRGB2HSB(RGB rgb, HSB* outHSB);
```
Any reason to use pointers here versus just a return value?

grp commented on FBTweak/FBColorUtils.h:
```diff
@@ -0,0 +1,80 @@
+ *  @param outRGB   The rgb color values
+ *  @param hsb The hsb color values
+ */
+extern void FBHSB2RGB(HSB hsb, RGB* outRGB);
```
Same for the rest of these.

grp commented on FBTweak/_FBColorComponentCell.m:
```diff
@@ -0,0 +1,169 @@
+  NSString *newString = [textField.text stringByReplacingCharactersInRange:range withString:string];
+
+  //first, check if the new string is numeric only. If not, return NO;
+  NSCharacterSet *characterSet = [[NSCharacterSet characterSetWithCharactersInString:@"0123456789,."] invertedSet];
```
I think `+[NSCharacterSet decimalDigitCharacterSet]` would work here? Perhaps combined with "." and "," as needed.

grp commented on FBTweak/_FBColorComponentCell.m:
```diff
@@ -0,0 +1,169 @@
+  _textField = [[UITextField alloc] init];
+  _textField.textAlignment = NSTextAlignmentCenter;
+  _textField.translatesAutoresizingMaskIntoConstraints = NO;
+  [_textField setKeyboardType:UIKeyboardTypeNumbersAndPunctuation];
```
nit: dot synatx

grp commented on FBTweak/_FBColorWheelView.m:
```diff
@@ -0,0 +1,182 @@
+  }
+}
+
+- (id)imageWithRGBAData:(CFDataRef)data width:(NSUInteger)width  height:(NSUInteger)height
```
Perhaps make this a function rather than a method, since it doesn't look like it uses any state from the object?

grp commented on FBTweak/_FBKeyboardManager.h:
```diff
@@ -0,0 +1,34 @@
+/**
+ @abstract Keyboard manager. Adjusts the content so that the target object remains visible.
+ */
+@interface _FBKeyboardManager : NSObject
```
Nice!

grp commented on FBTweak/_FBTweakTableViewCell.m:
```diff
@@ -8,8 +8,28 @@
 #import "_FBTweakTableViewCell.h"
 
+@interface UIImage (Utils)
++ (UIImage*)imageWithColor:(UIColor*)color size:(CGSize)size;
```
Perhaps make this a C function so it doesn't override any app-defined categories that might use the same name?

grp commented on FBTweak/_FBTweakColorViewController.m:
```diff
@@ -0,0 +1,119 @@
+  return FBColorFromHexString(value);
+}
+
+- (IBAction)_segmentControlDidChangeValue:(UISegmentedControl*)sender
```
nit: `IBAction` is unnecessary

grp commented on FBTweak/_FBColorComponentCell.m:
```diff
@@ -0,0 +1,169 @@
+@interface _FBColorComponentCell () <UITextFieldDelegate>
+{
+
+@private
```
nit: for ivars (in all the classes you added), you can move the `{ /* ivars */ }` block onto the `@implementation` and remove the `@private` (it's implicit)

grp commented:
Finally, any chance you could post a few screenshots? I'll download it and try it out but best to see it in context.

sgl0v commented:
Hello Grant,

Looks like I managed to fix all issues you mentioned.
* Replaced the hex-string color representation with UIColor;
* Removed the auto-layout constraints, added the `layoutSubviews` method implementations instead; 
* Minor changes according to the code review.

Yes, probably it would be better to start from the RGB support. 

Now it works in this way:
![Alt Text](https://www.dropbox.com/s/0ydxl48g9xkcui1/out.gif?raw=1)

grp commented:
It's merged! I did a little bit of cleanup on the documentation side, but otherwise looks awesome. Thanks again for making this!

alistra commented:
I waited for this for so long, is there any cocoapods release planned soon with this?

wjmelements commented on FBTweakTests/FBTweakInlineTestsMRR.m:
```diff
@@ -58,6 +59,9 @@ - (void)testValueTypes
   XCTAssertEqualObjects(testNSString, @"one", @"NSString %@", testNSString);
 
+  __attribute__((unused)) UIColor *testUIColor = FBTweakValue(@"UIColor", @"UIColor", @"UIColor", [UIColor redColor], [UIColor redColor]);
+  XCTAssertEqualObjects([UIColor redColor], [UIColor redColor], @"UIColor %@", testUIColor);
```
this doesn't test anything

wjmelements replied:
it should be 
`XCTAssertEqualObjects(testUIColor, [UIColor redColor], @"UIColor %@", testUIColor);`

wjmelements replied:
https://github.com/facebook/Tweaks/pull/127
